### PR TITLE
Development

### PR DESCRIPTION
**Pull request by imadu**
Breaking changes

### CODE REVIEW

**Comment by imadu** [comment]
I am creating an e-commerce API

1) In the DB folder, I tried to sync the database once so that I didn't have to write a new connection in each of the packages

2) The User and Product folders have both structs (models) and Handlers(basic controllers). Each has a method for basic CRUD operations. I know that the files are a bit dirty, but I wasn't sure of how to separate the Database actions from the controller actions. I'm open to suggestions.

3) Auth and orders are not yet complete, but they would most likely hold the same format as the aforementioned folders

File: categories/handler.go [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+	"github.com/imadu/e-commerce-api/db"
+	"github.com/imadu/e-commerce-api/util"
+	"github.com/labstack/echo"
+	"github.com/labstack/gommon/log"
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/bson/primitive"
+	"go.mongodb.org/mongo-driver/mongo/options"
+)
+
+var categoryCollection = db.Client.Database("cakes-and-cream-go").Collection("categories")
```
Comment (abiodun0) [inline categories/handler.go]: ***Nitpick***

I'm normally against hardcoding variables like this, you should have a configuration env files that are able to get your mongo database off the config and abstract it to a function as needed

File: categories/handler.go [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+// CreateNewCategory function
+func CreateNewCategory(c echo.Context) error {
+	r := new(Category)
```
Comment (abiodun0) [inline categories/handler.go]: ***Nitpick***

`r` is such a great name for a Category struct, can you name it to something more meaningful so that a reader doesn't have to spend a cognitive load following the function?

File: categories/handler.go [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+	if err := c.Bind(r); err != nil {
```
Comment (abiodun0) [inline categories/handler.go]: ***Excellent***

I like your use of early return, it follows with the step down principle

File: categories/handler.go [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+	result, _ := categoryCollection.InsertOne(ctx, r)
+	return util.SendSuccess(c, result.InsertedID)
+}
+
+// GetCategories Function
+func GetCategories(c echo.Context) error {
+	findOptions := options.Find()
```
Comment (abiodun0) [inline categories/handler.go]: ***Critical***

This 
``` 
findOptions := options.Find()
findOptions.SetLimit(10)
```
appears in many places in your app, I think they belong to the top of the app or inisde the main funciton

File: categories/handler.go [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+	findOptions.SetLimit(10)
+	var categories []*Category
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
```
Comment (abiodun0) [inline categories/handler.go]: ***Unsure/Maybe critical***

this is a bit dicey, why would you want to hardcode timeout on a database call? what if you are on a particularly slow server?

I think this should be entirely dependent on the response you get back from the database

File: categories/handler.go [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+var categoryCollection = db.Client.Database("cakes-and-cream-go").Collection("categories")
+
+// CreateNewCategory function
+func CreateNewCategory(c echo.Context) error {
+	r := new(Category)
+	if err := c.Bind(r); err != nil {
+		log.Errorf("Could not bind request to struct: %+v", err)
+		return util.SendError(c, "", "", "")
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
```
Comment (abiodun0) [inline categories/handler.go]: Same with my comment on withTimeout, I think timeout shouldn't be used with any IO based call

File: categories/handler.go [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+	for result.Next(ctx) {
+		var category Category
+		err := result.Decode(&category)
+		if err != nil {
+			log.Errorf("something went wrong", err)
+		}
+		categories = append(categories, &category)
+	}
+
+	if err := result.Err(); err != nil {
```
Comment (abiodun0) [inline categories/handler.go]: ***Critical***

Should this not be above the code and returning early? if there is ever an error?

File: categories/handler.go [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+	return util.SendData(c, categories)
+}
+
+// GetCategory function
+func GetCategory(c echo.Context) error {
+	id := c.Param("id")
+	var result Category
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+	err := categoryCollection.FindOne(ctx, id).Decode(&result)
```
Comment (abiodun0) [inline categories/handler.go]: ***Nitpick***
I like things being explicit, I would have preferred if you assign the result variable directly rather than passing by reference, but that can be subjective

File: categories/handler.go [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+	if err != nil {
+		c.Response().WriteHeader(http.StatusNotFound)
+		return util.SendError(c, "404", "Category do not exist", "failed")
+	}
+	return util.SendData(c, result)
+}
+
+//UpdateCategory function
+func UpdateCategory(c echo.Context) error {
```
Comment (abiodun0) [inline categories/handler.go]: ***Code format***

While reading down I feel there is a lot of repetition that could have been avoided using strategy pattern

In a way, what you are calling is the categoryCollection methods, so you could have a strategey that picks up whever one you are doing and do a switch as the need may be

Without needing to do a defer cancel and each and every one of them, you also get the parameters as needed, it might be in form of a 
```
type Parameters Struct
```

File: categories/struct.go [hunk 10]
```diff
@@ -0,0 +1,14 @@
+package categories
```
Comment (abiodun0) [inline categories/struct.go]: ***Code structure***

Since this is such a small application, I think that all the struct can be in one place

File: db/db.go [hunk 11]
```diff
@@ -3,19 +3,32 @@ package db
...
-	return client, &ctx
+//Client is the exported client that returns the database
+var Client *mongo.Client
+
+//InitDB exposes the database so we can pass it around in the other handlers
+func InitDB() *mongo.Client {
+	dbURI := os.Getenv("DB_URI")
+	once.Do(func() {
+		var err error
+		ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
```
Comment (abiodun0) [inline db/db.go:24]: I believe you can do without the Timeout here by just cancelling as needed rather than basing it on time

File: db/models/order.go [hunk 12]
```diff
@@ -1,13 +0,0 @@
-package models
```
Comment (abiodun0) [inline db/models/order.go:1]: Why was this deleted?

File: orders/struct.go [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+Denied const
+Failed const
+*/
+const (
+	Paid Payment = iota
+	Denied
+	Failed
+)
+
+func (p Payment) String() string {
```
Comment (abiodun0) [inline orders/struct.go]: what are you trying to achieve here, an enum?

File: products/handler.go [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+type Query struct {
+	Name     string
+	Category string
+	Limit    int64
+	Page     int64
+}
+
+var dbName = os.Getenv("DB_NAME")
+
+var productCollection = db.Client.Database(dbName).Collection("products")
```
Comment (abiodun0) [inline products/handler.go]: ***Critical***

Inconsistencies, this should be a function that returns the collection you are dealing with at that point in time, same with category collections

File: products/handler.go [hunk 15]
```diff
@@ -0,0 +1,131 @@
...
+
+func getName(name string) Product {
+	var result Product
+	filter := bson.D{primitive.E{Key: "name", Value: name}}
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+	err := productCollection.FindOne(ctx, filter).Decode(&result)
+	if err != nil {
+		log.Fatal(err)
```
Comment (abiodun0) [inline products/handler.go:17]: ***Critical***

You fail to return early here, that's going to affect this code moving forward

File: products/handler.go [hunk 16]
```diff
@@ -0,0 +1,131 @@
...
+
+}
+
+//GetProducts returns a list of products based on the query parameters
+func GetProducts(c echo.Context) error {
+	var q Query
+	q.Limit, _ = strconv.ParseInt(c.QueryParam("limit"), 10, 64)
+	q.Page, _ = strconv.ParseInt(c.QueryParam("page"), 10, 64)
+
+	findOptions := options.Find()
```
Comment (abiodun0) [inline products/handler.go]: ***Cricital***

This should be abstracted in its own function with it's own valid parameters, it's brittle and would break in the long runn

File: products/handler.go [hunk 17]
```diff
@@ -0,0 +1,131 @@
...
+	findOptions.SetLimit(q.Limit)
+	findOptions.SetSkip(q.Page * q.Limit)
+
+	var results []*Product
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+
+	cur, err := productCollection.Find(ctx, bson.D{{}}, findOptions)
```
Comment (abiodun0) [inline products/handler.go]: ***Critical/Naming***

You have inconsistent naming conventions, `curr` doesn't relate with anything here, perhaps use `result` as you've been doing?

File: products/handler.go [hunk 18]
```diff
@@ -0,0 +1,131 @@
...
+	if err := cur.Err(); err != nil {
+		log.Fatal(err)
+	}
+	cur.Close(ctx)
+	return util.SendData(c, results)
+
+}
+
+//Create creates a product
+func Create(c echo.Context) error {
```
Comment (abiodun0) [inline products/handler.go]: ***Critical***

What makes this function so different from the one on categories? I think it can be better even if you are setting some part of the columns to be unique

File: products/handler.go [hunk 19]
```diff
@@ -0,0 +1,131 @@
...
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	mod := mongo.IndexModel{
+		Keys: bson.M{
+			"name": -1,
+		}, Options: options.Index().SetUnique(true),
+	}
+
+	ind, err := productCollection.Indexes().CreateOne(ctx, mod)
+	if err != nil {
+		log.Fatal("Indexes().CreateOne() ERROR:", err)
```
Comment (abiodun0) [inline products/handler.go]: ***Nitpick***

If you had return early, you don't need an else

File: products/handler.go [hunk 20]
```diff
@@ -0,0 +1,131 @@
...
+		fmt.Println("CreateOne() type:", reflect.TypeOf(ind))
+	}
+
+	p := new(Product)
+	if err := c.Bind(p); err != nil {
+		log.Fatalf("Could not bind request to struct: %+v", err)
+		defer cancel()
+		return util.SendError(c, "500", "something went wrong", "failed")
+	}
+	defer cancel()
```
Comment (abiodun0) [inline products/handler.go]: ***Nitpick***

if you had this `defer cancel` at the top, there would be no need to repeat it twice

File: server.go [hunk 21]
```diff
@@ -1,16 +1,26 @@
...
 	"github.com/labstack/echo"
+	"github.com/subosito/gotenv"
 )
 
 func main() {
-	e := echo.New()
-	e.GET("/", func(c echo.Context) error {
+	err := gotenv.Load()
+	if err != nil {
+		log.Fatalf("Could not load env file: %+v", err)
```
Comment (abiodun0) [inline server.go:21]: You should return early, to avoid fatal errors, as this won't stop it from running the next line

File: users/handler.go [hunk 22]
```diff
@@ -0,0 +1,172 @@
...
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	mod := mongo.IndexModel{
+		Keys: bson.M{
+			"username": -1,
+		}, Options: options.Index().SetUnique(true),
+	}
+
+	ind, err := userCollection.Indexes().CreateOne(ctx, mod)
+	if err != nil {
+		log.Fatalf("Indexes().CreateOne() ERROR:", err)
```
Comment (abiodun0) [inline users/handler.go]: return early no need of else statement in here

File: users/handler.go [hunk 23]
```diff
@@ -0,0 +1,172 @@
...
+	if err := cur.Err(); err != nil {
+		log.Fatal(err)
+	}
+	cur.Close(ctx)
+	return util.SendData(c, results)
+
+}
+
+// CreateUser creates a new user
+func CreateUser(c echo.Context) error {
```
Comment (abiodun0) [inline users/handler.go]: Same comment as to when you are creating a product, and it seems some repetitions can be abstracted

File: users/handler.go [hunk 24]
```diff
@@ -0,0 +1,172 @@
...
+		fmt.Println("CreateOne() index:", ind)
+		fmt.Println("CreateOne() type:", reflect.TypeOf(ind))
+	}
+
+	hashedPassword, err := hashPassword(password)
+	if err != nil {
+		log.Fatalf("could not hash password", err)
+	}
+
+	u := new(User)
```
Comment (abiodun0) [inline users/handler.go]: ***Critical***

Use meaningful variable names

File: users/handler.go [hunk 25]
```diff
@@ -0,0 +1,172 @@
...
+	u.Password = hashedPassword
+	if err := c.Bind(u); err != nil {
+		log.Errorf("Could not bind request to struct: %+v", err)
+		defer cancel()
+		return util.SendError(c, "500", "something went wrong", "failed")
+	}
+	defer cancel()
```
Comment (abiodun0) [inline users/handler.go]: ***Nitpick***
Deferring at the top of the  if statement prevents repetition

File: users/handler.go [hunk 26]
```diff
@@ -0,0 +1,172 @@
...
+
+// UpdateUser updates the user
+func UpdateUser(c echo.Context) error {
+	id := c.Param("id")
+	body := new(User)
+	filter := bson.D{primitive.E{Key: "_id", Value: id}}
+	update := bson.D{primitive.E{Key: "firstName", Value: body.Firstname}, primitive.E{Key: "lastName", Value: body.Lastname}, primitive.E{Key: "email", Value: body.Email}, primitive.E{Key: "phoneNumber", Value: body.Phonenumber}}
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+	result, err := userCollection.UpdateOne(ctx, filter, update)
```
Comment (abiodun0) [inline users/handler.go]: ***Critical***

This is inconsistent with the pass by reference you've been using at one part of the app

File: util/util.go [hunk 27]
```diff
@@ -0,0 +1,50 @@
...
+	Status  string `json:"status"`
+}
+
+//ErrorResponse struct
+type ErrorResponse struct {
+	Error ErrorBody `json:"error"`
+}
+
+//SendSuccess utility function
+func SendSuccess(c echo.Context, data interface{}) error {
```
Comment (abiodun0) [inline util/util.go:28]: I think all this could have been abstracted into a sendMessage function

Everything could have been sent via data interface and use the status as needed

**Review by abiodun0 — COMMENTED** [review]
Good job!

Most of what I see are inconsistencies and use of interlocking conventions, and perhaps, an inapproriate use of withTimeout feature

It's a rule of thumb that softwares that depends on timeout for failures are generally to be avoided, the failure must be intentionally thrown by the IO source

File: products/service.go [hunk 28]
```diff
@@ -0,0 +1,121 @@
...
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/mongo/options"
+)
+
+var dbName = os.Getenv("DB_NAME")
+
+var productCollection = db.Client.Database(dbName).Collection("products")
+
+var ctx, cancel = context.WithTimeout(context.Background(), 10*time.Second)
```
Comment (abiodun0) [inline products/service.go:18]: I think you no longer need timeout in this case, because it would have already canceled before any of the function gets called
